// hexo-custom-autolinker/lib/filter.js
var Autolinker = require('autolinker');
var cheerio = require('cheerio');
var _ = require('lodash');
var urlparser = require('url');

var _defaultConfig = {
    replaceFn : function( autolinker, match ) {
        //console.log(match.getType());
        //console.log(match.getUrl());
        if (match.getType() == 'url') {
            var type = 'url';
            if (match.getUrl().indexOf('twitter.com') != -1) {
              var parsed = urlparser.parse(match.getUrl());
              var tweet_id = parsed['path'].split('/')[3];
              var type = 'url';

              return '<div class="twitter_wrap tweet" data-tweet-id="'+tweet_id+'"><a href="'+match.url+'" class="tweet_url">'+match.matchedText+'</a></div>';
            }
            if ((match.getUrl().indexOf('youtube.com') != -1) || (match.getUrl().indexOf('youtu.be') != -1) || (match.getUrl().indexOf('vimeo.com') != -1) || (match.getUrl().indexOf('vimeo.com') != -1) || (match.getUrl().indexOf('dai.ly') != -1)) {
                var parsed = urlparser.parse(match.getUrl());
                
                //console.log( "url: ", match.getUrl() );
                if (match.getUrl().indexOf('vimeo.com') != -1) {
                    console.log("Vimeo video found...");
                    var type = 'vimeo';
                    var id = parsed['path'].replace('/', '');
                }
                if (match.getUrl().indexOf('dai.ly') != -1) {
                    console.log("Daily Motion video found...");
                    var type = 'dailymotion';
                    var id = parsed['path'].replace('/', '');
                }

                if ((match.getUrl().indexOf('youtube.com') != -1) || (match.getUrl().indexOf('youtu.be') != -1)) {
                    console.log("Youtube video found...");
                    var type = 'youtube';
                    if (match.getUrl().indexOf('youtu.be') != -1) {
                        var id = parsed['path'].replace('/', '');
                    }
                    else {
                        var id = parsed['query'].replace('v=', '');
                    }
                }
                return ResponsiveVideo(id, type);
            }
        }
    },
    newWindow: true,
    stripPrefix: false,
    phone: false,
    twitter: false,
    hashtag: false
};

ResponsiveVideo = function(id, type) {
    if (type == null) {
        type = 'youtube';
    }
    var output  = '';
    output += '<div class="video_wrapper_outer">';
    output += '<div class="video_wrapper">';
    if (type == 'youtube') {
        output += '<iframe class="lazy" width="580" height="326" src="#" data-src="//www.youtube.com/embed/'+id+'" frameborder="0" allowfullscreen></iframe>';
    }
    if (type == 'vimeo') {
        output += '<iframe class="lazy" src="#" data-src="//player.vimeo.com/video/'+id+'?badge=0" width="500" height="281" frameborder="0" webkitallowfullscreen mozallowfullscreen allowfullscreen></iframe>';
    }

    //http://dai.ly/x1ub7t6
    if (type == 'dailymotion') {
        output += '<iframe class="lazy" src="#" data-src="//www.dailymotion.com/embed/video/'+id+'" width="480" height="270" frameborder="0" webkitallowfullscreen mozallowfullscreen allowfullscreen></iframe>';
    }
    output += '</div>';
    output += '</div>';
    return output;
}

module.exports = function(source) {
    
    this.config = this.config || {};

    var config = this.config.autolinker || {};

    config = _.extend({}, _defaultConfig, config);
    var $ = cheerio.load(source);
    $("img").each(function() {
      if (!$(this).hasClass('lazy-ignore')) {
       
        if ($(this).attr("src")) {
          $(this).attr("data-src", $(this).attr("src"));
          $(this).attr("src", "");
          $(this).attr('class', 'lazy');
        }

      }
    });
    
    
    source = $.html();
    source = Autolinker.link(source, config);
    
    return source;
};
